cardano-cli/src/Cardano/CLI/Shelley/Run/Governance.hs

```diff
@@ -0,0 +1,125 @@
+    GovernanceWriteUpdateError fp apiError ->
+      "Error writing shelley update proposal at: " <> textShow fp <> " Error: " <> renderApiError apiError
+    -- TODO: The equality check is still not working for empty update proposals.
+    GovernanceEmptyUpdateProposal ->
+      "Empty update proposals are not allowed"
+
+runGovernanceCmd :: GovernanceCmd -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceCmd (GovernanceMIRCertificate vKeys rewards out) = runGovernanceMIRCertificate vKeys rewards out
+runGovernanceCmd (GovernanceUpdateProposal out eNo genVKeys ppUp) = runGovernanceUpdateProposal out eNo genVKeys ppUp
+runGovernanceCmd cmd = liftIO $ putStrLn $ "TODO: runGovernanceCmd: " ++ show cmd
+
+runGovernanceMIRCertificate
+  :: [VerificationKeyFile]
+  -- ^ Stake verification keys
+  -> [ShelleyCoin]
+  -- ^ Reward amounts
+  -> OutputFile
+  -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceMIRCertificate vKeys rwdAmts (OutputFile oFp) = do
+    sCreds <- mapM readStakeKeyToCred vKeys
+
+    let mirCert = shelleyMIRCertificate $ mirMap sCreds rwdAmts
+
+    firstExceptT (GovernanceWriteMIRCertError oFp) . newExceptT $ writeCertificate oFp mirCert
+  where
+    readStakeKeyToCred :: VerificationKeyFile -> ExceptT ShelleyGovernanceError IO ShelleyCredentialStaking
+    readStakeKeyToCred (VerificationKeyFile stVKey) = do
+      StakingVerificationKeyShelley stakeVkey <-
+        firstExceptT (GovernanceReadStakeVerKeyError stVKey) . newExceptT $ readStakingVerificationKey stVKey
+      right . mkShelleyStakingCredential $ hashKey stakeVkey
+
+    mirMap :: [ShelleyCredentialStaking] -> [ShelleyCoin] -> Map.Map ShelleyCredentialStaking ShelleyCoin
+    mirMap stakeCreds rwds = Map.fromList $ zip stakeCreds rwds
+
+runGovernanceUpdateProposal
+  :: OutputFile
+  -> EpochNo
+  -> [VerificationKeyFile]
+  -- ^ Genesis verification keys
+  -> ShelleyPParamsUpdate
+  -> ExceptT ShelleyGovernanceError IO ()
+runGovernanceUpdateProposal (OutputFile upFile) eNo genVerKeyFiles upPprams' = do
+    upPprams <- checkForEmptyProposal $! upPprams'
+    genVKeys <- mapM
+                  (\(VerificationKeyFile fp) -> do
+                    gvk <- firstExceptT (GovernanceReadGenVerKeyError fp) $ readVerKey GenesisKey fp
+                    pure gvk
+                  )
+                  genVerKeyFiles
+    let genKeyHashes = map hashKey genVKeys
+        upProp = ShelleyUpdate $ createShelleyUpdateProposal eNo genKeyHashes upPprams
+    firstExceptT (GovernanceWriteUpdateError upFile)  . newExceptT $ writeUpdate upFile upProp
+  where
+    checkForEmptyProposal :: ShelleyPParamsUpdate -> ExceptT ShelleyGovernanceError IO ShelleyPParamsUpdate
+    checkForEmptyProposal sPParams
+      | sPParams == emptyPParamsUpdate = left GovernanceEmptyUpdateProposal
+      | otherwise = right sPParams
+
+-- TODO: Import from shelley ledger specs
+emptyPParamsUpdate :: ShelleyPParamsUpdate
+emptyPParamsUpdate =
```

Comment:
I'll raise a PR in `shelley-ledger-specs` to export `emptyPParamsUpdate` so we can use it (I'll address this in a follow-up PR).

Reply:
This needs to be double checked at the equality isn't working.

Reply:
This has been exported: https://github.com/input-output-hk/cardano-ledger-specs/pull/1498